src/library/vm/interaction_state.h

```diff
@@ -0,0 +1,242 @@
+        lean_assert(dynamic_cast<vm_State *>(to_external(o)));
+        return static_cast<vm_State *>(to_external(o))->m_val;
+    }
+
+    static vm_obj to_obj(State const & s) {
+        return mk_vm_external(new(get_vm_allocator().allocate(sizeof(vm_State))) vm_State(s));
+    }
+
+    static bool is_success(vm_obj const & o) {
+        return is_constructor(o) && cidx(o) == 0;
+    }
+
+    typedef std::tuple<format, optional<expr>, State> exception_info;
+
+    static optional<exception_info> is_exception(vm_state & S, vm_obj const & ex) {
+        if (is_constructor(ex) && cidx(ex) == 1 && !is_none(cfield(ex, 0))) {
+            vm_obj fmt = S.invoke(get_some_value(cfield(ex, 0)), mk_vm_unit());
+            optional<expr> ref;
+            if (!is_none(cfield(ex, 1)))
+                ref = to_expr(get_some_value(cfield(ex, 1)));
+            return optional<exception_info>(to_format(fmt), ref, to_State(cfield(ex, 2)));
+        } else {
+            return optional<exception_info>();
+        }
+    }
+
+    static bool is_silent_exception(vm_obj const & ex) {
+        return is_constructor(ex) && cidx(ex) == 1 && is_none(cfield(ex, 0));
+    }
+
+    static vm_obj mk_result(vm_obj const & a, vm_obj const & s) {
+        lean_assert(is_State(s));
+        return mk_vm_constructor(0, a, s);
+    }
+
+    static bool is_result_exception(vm_obj const & r) {
+        return is_constructor(r) && cidx(r) == 1;
+    }
+
+    static bool is_result_success(vm_obj const & r) {
+        return is_constructor(r) && cidx(r) == 0;
+    }
+
+    static vm_obj get_result_value(vm_obj const & r) {
+        lean_assert(is_result_success(r));
+        return cfield(r, 0);
+    }
+
+    static vm_obj get_result_state(vm_obj const & r) {
+        lean_assert(is_result_success(r));
+        return cfield(r, 1);
+    }
+
+    static vm_obj mk_success(vm_obj const & a, State const & s) {
+        return mk_vm_constructor(0, a, to_obj(s));
+    }
+
+    static vm_obj mk_success(State const & s) {
+        return mk_success(mk_vm_unit(), s);
+    }
+
+    static vm_obj mk_exception(vm_obj const & fn, State const & s) {
+        return mk_vm_constructor(1, mk_vm_some(fn), mk_vm_none(), to_obj(s));
+    }
+
+    static vm_obj mk_silent_exception(State const & s) {
+        return mk_vm_constructor(1, mk_vm_none(), mk_vm_none(), to_obj(s));
+    }
+
+    static vm_obj mk_exception(vm_obj const & fn, vm_obj const & ref, State const & s) {
+        return mk_vm_constructor(1, mk_vm_some(fn), ref, to_obj(s));
+    }
+
+    static vm_obj mk_exception(throwable const & ex, State const & s) {
+        vm_obj _ex = lean::to_obj(ex);
+        vm_obj fn = mk_vm_closure(get_throwable_to_format_fun_idx(), 1, &_ex);
+        optional<expr> ref;
+        if (auto kex = dynamic_cast<ext_exception const *>(&ex))
+            ref = kex->get_main_expr();
+        else if (auto fex = dynamic_cast<formatted_exception const *>(&ex))
+            ref = fex->get_main_expr();
```

Comment:
@leodemoura This code should now also handle `parser_error`s in order to propagate error positions. But I don't see how to do that since those exceptions contain only a position, not an `expr`. We could further abstract over the type of information stored in `interaction_monad.result` (`pos` for parser, `expr` for tactic), but I don't think we actually do anything with the `expr` right now apart from querying its position. Should we switch to just
```
| exception {} : option (unit → format) → option pos /- currently expr -/ → state → interaction_monad.result
```
?